File: app/src/main/java/com/micewine/emu/core/RatPackageManager.java
package com.micewine.emu.core;

import static com.micewine.emu.activities.GeneralSettingsActivity.SELECTED_BOX64;
import static com.micewine.emu.activities.GeneralSettingsActivity.SELECTED_CORE;
import static com.micewine.emu.activities.GeneralSettingsActivity.SELECTED_VULKAN_DRIVER;
import static com.micewine.emu.activities.MainActivity.appRootDir;
import static com.micewine.emu.activities.MainActivity.gson;
import static com.micewine.emu.activities.MainActivity.preferences;
import static com.micewine.emu.activities.MainActivity.ratPackagesDir;
import static com.micewine.emu.adapters.AdapterRatPackage.BOX64;
import static com.micewine.emu.adapters.AdapterRatPackage.CORE;
import static com.micewine.emu.adapters.AdapterRatPackage.DXVK;
import static com.micewine.emu.adapters.AdapterRatPackage.VKD3D;
import static com.micewine.emu.adapters.AdapterRatPackage.VK_DRIVER;
import static com.micewine.emu.adapters.AdapterRatPackage.WINE;
import static com.micewine.emu.adapters.AdapterRatPackage.WINED3D;
import static com.micewine.emu.core.ShellLoader.runCommand;
import static com.micewine.emu.core.TarUtils.getFileStreamFromTarXZ;
import static com.micewine.emu.core.TarUtils.isXZ;
import static com.micewine.emu.core.TarUtils.untar;
import static com.micewine.emu.utils.FileUtils.deleteDirectoryRecursively;

import static java.util.UUID.randomUUID;

import android.content.SharedPreferences;

import com.micewine.emu.fragments.SetupFragment;

import net.lingala.zip4j.ZipFile;
import net.lingala.zip4j.io.inputstream.ZipInputStream;
import net.lingala.zip4j.model.FileHeader;
import net.lingala.zip4j.progress.ProgressMonitor;

import java.io.BufferedReader;
import java.io.ByteArrayInputStream;
import java.io.ByteArrayOutputStream;
import java.io.File;
import java.io.FileWriter;
import java.io.IOException;
import java.io.InputStream;
import java.io.InputStreamReader;
import java.nio.file.Files;
import java.util.ArrayList;
import java.util.List;
import java.util.Set;
import java.util.stream.Collectors;

public class RatPackageManager {
    public static void installRat(RatPackage ratPackage, SetupFragment.ProgressCallback callback) {
        callback.setProgressBarIndeterminate(false);
        callback.onProgressChanged(0);

        String extractDir = ratPackagesDir + "/" + ratPackage.category + "-" + randomUUID();
        if (!new File(extractDir).mkdirs()) return;

        if (ratPackage.isTarXZRat) {
            try {
                untar(ratPackage.ratFile.getPath(), extractDir, callback);
            } catch (IOException e) {
                return;
            }
        } else {
            try (ZipFile ratFile = new ZipFile(ratPackage.ratFile)) {
                ratFile.setRunInThread(true);
                ratFile.extractAll(extractDir);

                while (!ratFile.getProgressMonitor().getState().equals(ProgressMonitor.State.READY)) {
                    callback.onProgressChanged(ratFile.getProgressMonitor().getPercentDone());
                    Thread.sleep(125);
                }
            } catch (Exception ignored) {
                return;
            }
        }

        callback.onProgressChanged(0);

        runCommand("chmod -R 700 " + extractDir, false);
        runCommand("sh " + extractDir + "/makeSymlinks.sh " + extractDir + "/", false);

        new File(extractDir, "makeSymlinks.sh").delete();

        switch (ratPackage.category) {
            case "Core" -> {
                if (preferences == null) return;
                if (preferences.getString(SELECTED_CORE, "").isEmpty()) {
                    SharedPreferences.Editor editor = preferences.edit();
                    editor.putString(SELECTED_CORE, new File(extractDir).getName());
                    editor.apply();
                }
            }
            case "Box64" -> {
                if (preferences == null) return;
                if (preferences.getString(SELECTED_BOX64, "").isEmpty()) {
                    SharedPreferences.Editor editor = preferences.edit();
                    editor.putString(SELECTED_BOX64, new File(extractDir).getName());
                    editor.apply();
                }
            }
            case "VulkanDriver", "AdrenoTools" -> {
                if (preferences == null) return;
                if (preferences.getString(SELECTED_VULKAN_DRIVER, "").isEmpty()) {
                    SharedPreferences.Editor editor = preferences.edit();
                    editor.putString(SELECTED_VULKAN_DRIVER, new File(extractDir).getName());
                    editor.apply();
                }

                File pkgHeader = new File(extractDir, "pkg-header");

                try (FileWriter writer = new FileWriter(pkgHeader)) {
                    writer.write(
                            "name=" + ratPackage.name + "\n" +
                            "category=" + ratPackage.category + "\n" +
                            "version=" + ratPackage.version + "\n" +
                            "architecture=" + ratPackage.architecture + "\n" +
                            "vkDriverLib=" + extractDir + "/files/usr/lib/" + ratPackage.driverLib + "\n"
                    );
                } catch (IOException ignored) {
                }
            }
        }

        setPackageExternal(extractDir);
    }

    public static void deleteRatPackageById(String id) {
        RatPackage ratPackage = getPackageById(id);

        if (ratPackage == null) return;
        if (ratPackage.isUserInstalled) {
            deleteDirectoryRecursively(new File(ratPackagesDir, id).toPath());
        }
    }

    public static boolean haveAnyPackageByCategory(int category) {
        return !listRatPackages(getRatCategoryString(category)).isEmpty();
    }

    public static List<RatPackage> listRatPackages(int category) {
        return listRatPackages(getRatCategoryString(category), getRatCategoryString(category));
    }

    public static List<RatPackage> listRatPackages(int category, int anotherCategory) {
        return listRatPackages(getRatCategoryString(category), getRatCategoryString(anotherCategory));
    }

    public static List<RatPackage> listRatPackages(String category) {
        return listRatPackages(category, category);
    }

    public static List<RatPackage> listRatPackages(String category, String anotherCategory) {
        ArrayList<RatPackage> packagesList = new ArrayList<>();

        File packagesDir = new File(appRootDir, "packages");
        if (!packagesDir.exists()) return packagesList;

        File[] files = packagesDir.listFiles();
        if (files == null) return packagesList;

        for (File file : files) {
            if (file.isDirectory() && ((file.getName().startsWith(category + "-") || file.getName().startsWith(anotherCategory + "-")) || category.isEmpty() || anotherCategory.isEmpty())) {
                File pkgHeader = new File(file, "pkg-header");
                if (!pkgHeader.exists()) continue;

                List<String> lines;

                try {
                    lines = Files.readAllLines(pkgHeader.toPath());
                } catch (IOException ignored) {
                    lines = null;
                }

                if (lines == null || lines.size() < 5) continue;

                String pkgName = lines.get(0).substring(lines.get(0).indexOf("=") + 1);

                if (file.getName().startsWith("AdrenoToolsDriver-")) {
                    pkgName += " (AdrenoTools)";
                }

                String pkgCategory = lines.get(1).substring(lines.get(1).indexOf("=") + 1);
                String pkgVersion = lines.get(2).substring(lines.get(2).indexOf("=") + 1);
                String pkgDriverLib = lines.get(4).substring(lines.get(4).indexOf("=") + 1);

                packagesList.add(
                        new RatPackage(pkgName, pkgCategory, pkgVersion, pkgDriverLib, file.getName(), new File(file, "pkg-external").exists())
                );
            }
        }

        return packagesList;
    }

    public static List<String> listRatPackagesId(int category) {
        return listRatPackagesId(getRatCategoryString(category), getRatCategoryString(category));
    }

    public static List<String> listRatPackagesId(int category, int anotherCategory) {
        return listRatPackagesId(getRatCategoryString(category), getRatCategoryString(anotherCategory));
    }

    public static List<String> listRatPackagesId(String type) {
        return listRatPackagesId(type, type);
    }

    public static List<String> listRatPackagesId(String type, String anotherType) {
        ArrayList<String> packagesList = new ArrayList<>();

        File packagesDir = new File(appRootDir, "packages");
        if (!packagesDir.exists()) return packagesList;

        File[] files = packagesDir.listFiles();
        if (files == null) return packagesList;

        for (File file : files) {
            if (file.isDirectory() && (file.getName().startsWith(type + "-") || file.getName().startsWith(anotherType + "-"))) {
                packagesList.add(file.getName());
            }
        }

        return packagesList;
    }

    public static RatPackage getPackageById(String id) {
        File pkgHeader = new File(ratPackagesDir, id + "/pkg-header");
        if (pkgHeader.exists() && id != null) {
            List<String> lines;

            try {
                lines = Files.readAllLines(pkgHeader.toPath());
            } catch (IOException ignored) {
                lines = null;
            }

            if (lines == null || lines.size() < 5) return null;

            String pkgName = lines.get(0).substring(lines.get(0).indexOf("=") + 1);
            String pkgCategory = lines.get(1).substring(lines.get(1).indexOf("=") + 1);
            String pkgVersion = lines.get(2).substring(lines.get(2).indexOf("=") + 1);
            String pkgDriverLib = lines.get(4).substring(lines.get(4).indexOf("=") + 1);

            if (pkgCategory.startsWith("AdrenoToolsDriver-")) {
                pkgName += " (AdrenoTools)";
            }

            return new RatPackage(pkgName, pkgCategory, pkgVersion, pkgDriverLib, id, new File(ratPackagesDir, id + "/pkg-external").exists());
        }

        return null;
    }

    public static String getPackageNameVersionById(String id) {
        File pkgHeader = new File(ratPackagesDir, id + "/pkg-header");
        if (pkgHeader.exists() && id != null) {
            try {
                List<String> lines = Files.readAllLines(pkgHeader.toPath());
                return lines.get(0).substring(lines.get(0).indexOf("=") + 1) + " " + lines.get(2).substring(lines.get(2).indexOf("=") + 1);
            } catch (IOException ignored) {
            }
        }

        return null;
    }

    public static boolean checkPackageInstalled(String name, String category, String version) {
        List<RatPackage> ratPackageList = listRatPackages("", "");

        for (RatPackage ratPackage : ratPackageList) {
            if (ratPackage.name.equals(name) && ratPackage.category.equals(category) && ratPackage.version.equals(version)) {
                return true;
            }
        }

        return false;
    }

    public static void installADToolsDriver(AdrenoToolsPackage adrenoToolsPackage, SetupFragment.ProgressCallback callback) {
        callback.setProgressBarIndeterminate(false);
        callback.onProgressChanged(0);

        ZipFile adrenoToolsFile = adrenoToolsPackage.adrenoToolsFile;

        String extractDir = ratPackagesDir + "/AdrenoToolsDriver-" + randomUUID();

        try {
            adrenoToolsFile.setRunInThread(true);
            adrenoToolsFile.extractAll(extractDir);

            while (!adrenoToolsFile.getProgressMonitor().getState().equals(ProgressMonitor.State.READY)) {
                callback.onProgressChanged(adrenoToolsFile.getProgressMonitor().getPercentDone());
                Thread.sleep(125);
            }
        } catch (Exception ignored) {
            return;
        }

        runCommand("chmod -R 700 " + extractDir, false);

        String driverLibPath = extractDir + "/" + adrenoToolsPackage.driverLib;

        File pkgHeader = new File(extractDir, "pkg-header");

        try (FileWriter writer = new FileWriter(pkgHeader)) {
            writer.write("name=" + adrenoToolsPackage.name + "\ncategory=AdrenoToolsDriver\nversion=" + adrenoToolsPackage.version + "\narchitecture=aarch64\nvkDriverLib=" + driverLibPath + "\n");
        } catch (IOException ignored) {
        }

        setPackageExternal(extractDir);
    }

    private static void setPackageExternal(String extractDir) {
        try (FileWriter writer = new FileWriter(new File(extractDir, "pkg-external"))) {
            writer.write("");
        } catch (IOException ignored) {
        }
    }

    public static class AdrenoToolsPackage {
        String name;
        String version;
        String description;
        String driverLib;
        String author;
        ZipFile adrenoToolsFile;

        public AdrenoToolsPackage(String path) {
            adrenoToolsFile = new ZipFile(path);

            try {
                if (!adrenoToolsFile.isValidZipFile()) return;

                FileHeader metaHeader = adrenoToolsFile.getFileHeader("meta.json");
                ZipInputStream inputStream = adrenoToolsFile.getInputStream(metaHeader);

                String json;

                try (BufferedReader reader = new BufferedReader(new InputStreamReader(inputStream))) {
                    json = reader.lines().collect(Collectors.joining("\n"));
                    AdrenoToolsMetaInfo meta = gson.fromJson(json, AdrenoToolsMetaInfo.class);

                    name = meta.name;
                    version = meta.driverVersion;
                    description = meta.description;
                    driverLib = meta.libraryName;
                    author = meta.author;
                }
            } catch (IOException ignored) {
            }
        }

        public String getName() {
            return name;
        }

        public String getVersion() {
            return version;
        }

        public String getDescription() {
            return description;
        }
    }

    public static class AdrenoToolsMetaInfo {
        public String name;
        public String description;
        public String author;
        public String driverVersion;
        public String libraryName;
    }

    public static class RatPackage {
        String name;
        String category;
        String version;
        String architecture;
        String driverLib;
        boolean isUserInstalled;
        boolean isTarXZRat;
        File ratFile;
        String folderName;

        public RatPackage(String ratPath) {
            ratFile = new File(ratPath);
            InputStream inputStream = null;

            isTarXZRat = isXZ(ratPath);

            if (isTarXZRat) {
                inputStream = getFileStreamFromTarXZ(ratPath, "pkg-header");
            } else {
                try (ZipFile zipFile = new ZipFile(ratFile)) {
                    if (!zipFile.isValidZipFile() && !zipFile.isValidZipFile()) return;

                    FileHeader ratHeader = zipFile.getFileHeader("pkg-header");

                    try (InputStream is = zipFile.getInputStream(ratHeader)) {
                        ByteArrayOutputStream byteArrayOutputStream = new ByteArrayOutputStream();
                        byte[] buffer = new byte[8196];
                        int len;
                        while ((len = is.read(buffer)) != -1) {
                            byteArrayOutputStream.write(buffer, 0, len);
                        }
                        inputStream = new ByteArrayInputStream(byteArrayOutputStream.toByteArray());
                    }
                } catch (IOException ignored) {
                }
            }

            if (inputStream == null) return;

            try (BufferedReader reader = new BufferedReader(new InputStreamReader(inputStream))) {
                List<String> lines = reader.lines().collect(Collectors.toList());
                if (lines.size() < 5) return;

                name = lines.get(0).substring(lines.get(0).indexOf("=") + 1);
                category = lines.get(1).substring(lines.get(1).indexOf("=") + 1);
                version = lines.get(2).substring(lines.get(2).indexOf("=") + 1);
                architecture = lines.get(3).substring(lines.get(3).indexOf("=") + 1);
                driverLib = lines.get(4).substring(lines.get(4).indexOf("=") + 1);
            } catch (IOException ignored) {
            }
        }

        public RatPackage(String name, String category, String version, String driverLib, String folderName, boolean isUserInstalled) {
            this.name = name;
            this.category = category;
            this.version = version;
            this.driverLib = driverLib;
            this.folderName = folderName;
            this.isUserInstalled = isUserInstalled;
        }

        public String getName() {
            return name;
        }

        public String getArchitecture() {
            return architecture;
        }

        public String getFolderName() {
            return folderName;
        }

        public String getDriverLib() {
            return driverLib;
        }

        public String getVersion() {
            return version;
        }

        public String getCategory() {
            return category;
        }

        public boolean getIsUserInstalled() {
            return isUserInstalled;
        }
    }

    public static Set<String> installablePackagesCategories = Set.of("Core", "VulkanDriver", "Box64", "Wine", "DXVK", "WineD3D", "VKD3D");

    public static int getRatCategoryId(String category) {
        return switch (category) {
            case "Core" -> CORE;
            case "VulkanDriver" -> VK_DRIVER;
            case "Box64" -> BOX64;
            case "Wine" -> WINE;
            case "DXVK" -> DXVK;
            case "WineD3D" -> WINED3D;
            case "VKD3D" -> VKD3D;
            default -> -1;
        };
    }

    public static String getRatCategoryString(int category) {
        return switch (category) {
            case CORE -> "Core";
            case VK_DRIVER -> "VulkanDriver";
            case BOX64 -> "Box64";
            case WINE -> "Wine";
            case DXVK -> "DXVK";
            case WINED3D -> "WineD3D";
            case VKD3D -> "VKD3D";
            default -> "";
        };
    }
}

File: app/src/main/java/com/micewine/emu/core/RatPackageManager.kt
package com.micewine.emu.core

import android.annotation.SuppressLint
import android.content.Context
import androidx.preference.PreferenceManager
import com.google.gson.Gson
import com.micewine.emu.R
import com.micewine.emu.activities.GeneralSettingsActivity.Companion.SELECTED_BOX64
import com.micewine.emu.activities.MainActivity.Companion.appRootDir
import com.micewine.emu.activities.MainActivity.Companion.ratPackagesDir
import com.micewine.emu.core.ShellLoader.runCommand
import com.micewine.emu.fragments.SetupFragment.Companion.dialogTitleText
import com.micewine.emu.fragments.SetupFragment.Companion.progressBarIsIndeterminate
import com.micewine.emu.fragments.SetupFragment.Companion.progressBarValue
import net.lingala.zip4j.ZipFile
import net.lingala.zip4j.progress.ProgressMonitor
import java.io.File

object RatPackageManager {
    @SuppressLint("SetTextI18n")
    fun installRat(ratPackage: RatPackage, context: Context) {
        val preferences = PreferenceManager.getDefaultSharedPreferences(context)

        progressBarIsIndeterminate = false

        var extractDir = appRootDir.parent

        ratPackage.ratFile.use { ratFile ->
            ratFile?.isRunInThread = true

            if (ratPackage.category == "rootfs") {
                installingRootFS = true
            } else {
                extractDir = "$ratPackagesDir/${ratPackage.category}-${java.util.UUID.randomUUID()}"
                File(extractDir!!).mkdirs()
            }

            ratFile?.extractAll(extractDir)

            while (!ratFile?.progressMonitor?.state?.equals(ProgressMonitor.State.READY)!!) {
                progressBarValue = ratFile.progressMonitor.percentDone

                Thread.sleep(100)
            }
        }

        progressBarValue = 0

        runCommand("chmod -R 700 $extractDir")
        runCommand("sh $extractDir/makeSymlinks.sh").also {
            File("$extractDir/makeSymlinks.sh").delete()
        }

        when (ratPackage.category) {
            "rootfs" -> {
                File("$extractDir/pkg-header").renameTo(File("$ratPackagesDir/rootfs-pkg-header"))

                val adrenoToolsFolder = File("$extractDir/adrenoTools")
                val vulkanDriversFolder = File("$extractDir/vulkanDrivers")
                val box64Folder = File("$extractDir/box64")
                val wineFolder = File("$extractDir/wine")

                File("$appRootDir/wine-utils/DXVK").listFiles()?.forEach {
                    val dxvkDir = File("$ratPackagesDir/DXVK-${java.util.UUID.randomUUID()}")
                    dxvkDir.mkdirs()
                    val dxvkFilesDir = File("$dxvkDir/files")
                    dxvkFilesDir.mkdirs()
                    val dxvkVersion = it.name.substringAfter("-")

                    it.renameTo(dxvkFilesDir)

                    File("$dxvkDir/pkg-header").writeText("name=DXVK\ncategory=DXVK\nversion=$dxvkVersion\narchitecture=any\nvkDriverLib=\n")
                }

                File("$appRootDir/wine-utils/WineD3D").listFiles()?.forEach {
                    val wineD3DDir = File("$ratPackagesDir/WineD3D-${java.util.UUID.randomUUID()}")
                    wineD3DDir.mkdirs()
                    val wineD3DFilesFir = File("$wineD3DDir/files")
                    wineD3DFilesFir.mkdirs()
                    val wineD3DVersion = it.name.substringAfter("-")

                    it.renameTo(wineD3DFilesFir)

                    File("$wineD3DDir/pkg-header").writeText("name=WineD3D\ncategory=WineD3D\nversion=$wineD3DVersion\narchitecture=any\nvkDriverLib=\n")
                }

                File("$appRootDir/wine-utils/VKD3D").listFiles()?.forEach {
                    val vkd3dDir = File("$ratPackagesDir/VKD3D-${java.util.UUID.randomUUID()}")
                    vkd3dDir.mkdirs()
                    val vkd3dFilesDir = File("$vkd3dDir/files")
                    vkd3dFilesDir.mkdirs()
                    val vkd3dVersion = it.name.substringAfter("-")

                    it.renameTo(vkd3dFilesDir)

                    File("$vkd3dDir/pkg-header").writeText("name=VKD3D\ncategory=VKD3D\nversion=$vkd3dVersion\narchitecture=any\nvkDriverLib=\n")
                }

                dialogTitleText = context.getString(R.string.installing_drivers)

                if (vulkanDriversFolder.exists()) {
                    vulkanDriversFolder.listFiles()?.sorted()?.forEach { ratFile ->
                        installRat(RatPackage(ratFile.path), context)
                    }

                    vulkanDriversFolder.deleteRecursively()
                }

                if (adrenoToolsFolder.exists()) {
                    adrenoToolsFolder.listFiles()?.sorted()?.forEach { ratFile ->
                        installRat(RatPackage(ratFile.path), context)
                    }

                    adrenoToolsFolder.deleteRecursively()
                }

                dialogTitleText = context.getString(R.string.installing_box64)

                if (box64Folder.exists()) {
                    box64Folder.listFiles()?.sorted()?.forEach { ratFile ->
                        installRat(RatPackage(ratFile.path), context)
                    }

                    box64Folder.deleteRecursively()
                }

                dialogTitleText = context.getString(R.string.installing_wine)

                if (wineFolder.exists()) {
                    wineFolder.listFiles()?.sorted()?.forEach { ratFile ->
                        installRat(RatPackage(ratFile.path), context)
                    }

                    wineFolder.deleteRecursively()
                }

                installingRootFS = false
            }
            "Box64" -> {
                preferences.apply {
                    if (getString(SELECTED_BOX64, "") == "") {
                        edit().apply {
                            putString(SELECTED_BOX64, File(extractDir!!).name)
                            apply()
                        }
                    }
                }

                File("$extractDir/pkg-header").writeText("name=${ratPackage.name}\ncategory=${ratPackage.category}\nversion=${ratPackage.version}\narchitecture=${ratPackage.architecture}\nvkDriverLib=\n")

                if (!installingRootFS) {
                    File("$extractDir/pkg-external").writeText("")
                }
            }
            "VulkanDriver", "AdrenoTools" -> {
                val driverLibPath = "$extractDir/files/usr/lib/${ratPackage.driverLib}"

                File("$extractDir/pkg-header").writeText("name=${ratPackage.name}\ncategory=${ratPackage.category}\nversion=${ratPackage.version}\narchitecture=${ratPackage.architecture}\nvkDriverLib=$driverLibPath\n")

                if (!installingRootFS) {
                    File("$extractDir/pkg-external").writeText("")
                }
            }
            "Wine", "DXVK", "WineD3D", "VKD3D" -> {
                File("$extractDir/pkg-header").writeText("name=${ratPackage.name}\ncategory=${ratPackage.category}\nversion=${ratPackage.version}\narchitecture=${ratPackage.architecture}\nvkDriverLib=\n")

                if (!installingRootFS) {
                    File("$extractDir/pkg-external").writeText("")
                }
            }
        }
    }

    fun listRatPackages(type: String = "", anotherType: String = "?"): List<RatPackage> {
        val packagesList: MutableList<RatPackage> = mutableListOf()

        File("$appRootDir/packages").listFiles()?.forEach { file ->
            if (file.isDirectory && (file.name.startsWith(type) || file.name.startsWith(anotherType))) {
                val lines = File("$file/pkg-header").readLines()

                var pkgName = lines[0].substringAfter("=")

                if (file.name.startsWith("AdrenoToolsDriver-")) {
                    pkgName += " (AdrenoTools)"
                }

                val pkgCategory = lines[1].substringAfter("=")
                val pkgVersion = lines[2].substringAfter("=")
                val pkgDriverLib = lines[3].substringAfter("=")

                packagesList.add(
                    RatPackage().apply {
                        name = pkgName
                        category = pkgCategory
                        version = pkgVersion
                        driverLib = pkgDriverLib
                        folderName = file.name

                        isUserInstalled = File("$file/pkg-external").exists()
                    }
                )
            }
        }

        return packagesList
    }

    fun listRatPackagesId(type: String = "", anotherType: String = "?"): List<String> {
        val packagesIdList: MutableList<String> = mutableListOf()

        File("$appRootDir/packages").listFiles()?.forEach { file ->
            if (file.isDirectory && (file.name.startsWith(type) || file.name.startsWith(anotherType))) {
                packagesIdList.add(file.name)
            }
        }

        return packagesIdList
    }

    fun getPackageNameVersionById(id: String): String {
        val lines = File("$ratPackagesDir/$id/pkg-header").readLines()

        return lines[0].substringAfter("=") + " " + lines[2].substringAfter("=")
    }

    fun installADToolsDriver(adrenoToolsPackage: AdrenoToolsPackage) {
        progressBarIsIndeterminate = false

        var extractDir: String

        adrenoToolsPackage.adrenoToolsFile.use {
            it.isRunInThread = true

            extractDir = "$ratPackagesDir/AdrenoToolsDriver-${java.util.UUID.randomUUID()}"

            it.extractAll(extractDir)

            while (!it.progressMonitor.state.equals(ProgressMonitor.State.READY)) {
                progressBarValue = it.progressMonitor.percentDone

                Thread.sleep(100)
            }
        }

        progressBarValue = 0

        runCommand("chmod -R 700 $extractDir")

        val driverLibPath = "$extractDir/${adrenoToolsPackage.driverLib}"

        File("$extractDir/pkg-header").writeText("name=${adrenoToolsPackage.name}\ncategory=AdrenoToolsDriver\nversion=${adrenoToolsPackage.version}\narchitecture=aarch64\nvkDriverLib=$driverLibPath\n")

        if (!installingRootFS) {
            File("$extractDir/pkg-external").writeText("")
        }
    }

    class AdrenoToolsPackage(path: String) {
        var name: String? = null
        var version: String? = null
        var description: String? = null
        var driverLib: String? = null
        var author: String? = null
        var adrenoToolsFile: ZipFile = ZipFile(path)

        init {
            adrenoToolsFile.getInputStream(adrenoToolsFile.getFileHeader("meta.json")).use { inputStream ->
                val json = inputStream.reader().readLines().joinToString("\n")
                val meta = Gson().fromJson(json, AdrenoToolsMetaInfo::class.java)

                name = meta.name
                version = meta.driverVersion
                description = meta.description
                driverLib = meta.libraryName
                author = meta.author
            }
        }
    }

    data class AdrenoToolsMetaInfo(
        val name: String,
        val description: String,
        val author: String,
        val driverVersion: String,
        val libraryName: String
    )

    class RatPackage(ratPath: String? = null) {
        var name: String? = null
        var category: String? = null
        var version: String? = null
        var architecture: String? = null
        var driverLib: String? = null
        var isUserInstalled: Boolean? = null
        var folderName: String? = null
        var ratFile: ZipFile? = null

        init {
            if (ratPath != null) {
                ratFile = ZipFile(ratPath)
                ratFile?.getInputStream(ratFile?.getFileHeader("pkg-header")).use { inputStream ->
                    val lines = inputStream?.reader()?.readLines()!!

                    name = lines[0].substringAfter("=")
                    category = lines[1].substringAfter("=")
                    version = lines[2].substringAfter("=")
                    architecture = lines[3].substringAfter("=")
                    driverLib = lines[4].substringAfter("=")
                }
            }
        }
    }

    private var installingRootFS: Boolean = false

    val installablePackagesCategories = setOf("VulkanDriver", "Box64", "Wine", "DXVK", "WineD3D", "VKD3D")
}
